src/python/pants/backend/python/rules/python_create_binary.py

```diff
@@ -2,28 +2,25 @@
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
 from dataclasses import dataclass
-from typing import Optional
+from typing import ClassVar, Optional
 
 from pants.backend.python.rules.pex import Pex
 from pants.backend.python.rules.pex_from_target_closure import CreatePexFromTargetClosure
 from pants.backend.python.rules.targets import EntryPoint, PythonBinarySources
 from pants.backend.python.targets.python_binary import PythonBinary
 from pants.build_graph.address import Address
 from pants.engine.addressable import Addresses
-from pants.engine.legacy.structs import PythonBinaryAdaptor
 from pants.engine.rules import UnionRule, rule
 from pants.engine.selectors import Get
-from pants.engine.target import Target, WrappedTarget
-from pants.rules.core.binary import BinaryTarget, CreatedBinary
+from pants.engine.target import Target
+from pants.rules.core.binary import BinaryImplementation, CreatedBinary
 from pants.rules.core.strip_source_roots import SourceRootStrippedSources, StripSourcesFieldRequest
 
 
-# TODO: consider replacing this with sugar like `SelectFields(EntryPoint, PythonBinarySources)` so
-#  that the rule would request that instead of this dataclass. Note that this syntax must support
-#  both optional_fields (see the below TODO) and opt-out `SentinelField`s
-#  (see https://github.com/pantsbuild/pants/pull/9316#issuecomment-600152573).
 @dataclass(frozen=True)
-class PythonBinaryFields:
+class PythonBinaryImplementation(BinaryImplementation):
```

Comment:
Keep reading the PR, especially `binary.py`, for why I found `BinaryImplementation` to be much more natural than `BinaryFields`. 

As mentioned in the PR description, another reason I think` BinaryImplementation` is better than `BinaryFields` is that it fits better with `test.py`, `fmt.py`, and `lint.py`, which use `TestRunner`, `Formatter`, and `Linter`, respectively, for this same construct.

Reply:
Both this form and the prior seem like more boilerplate than needed. Could non `Optional` fields be used as the signal to an external tester / creator? Given a `@dataclass`, the meaning of `Optional` and a `Target` instance, it seems you could have one piece of code that know if that `@dataclass` can be created from the Target and can then create an instance of the `@dataclass` if so.

Reply:
>  Given a @dataclass, the meaning of Optional and a Target instance, it seems you could have one piece of code that knows if that @dataclass can be created from the Target and can then create an instance of the @dataclass if so.

I don't follow. Do you mean factoring this out into some abstraction that removes most the boilerplate here? Or do you mean a more fundamental redesign?

If the former, almost certainly, yes, we can reduce boilerplate by refactoring this. But, I'm intentionally punting on boilerplate reduction until we finish porting all of V2 Python to the Target API so that we better understand all the usages we need to handle.

Reply:
OK. The change from PythonBinaryFields / PythonBinaryImplementation just looks like a non-obvious win and the refactor look like the more obvious win just reading this PR. Happy to ride out the thrash to get more data.

Reply:
> The change from PythonBinaryFields / PythonBinaryImplementation just looks like a non-obvious win

My bad with a confusing PR comment at the start of this thread. There are two things going on with this PR in regards to `PythonBinaryFields`:

1. Rename `PythonBinaryFields` to `PythonBinaryImplementation`. This is solely a name change to reflect what ended up being far more natural in real-world usage, e.g. what sounded better in the error message in the PR description.
   * This is what the original comment in this thread was referring to.
2. Have `PythonBinaryImplementation` subclass `BinaryImplementation`, whereas previously it was a stand-alone dataclass.
   * This removes some boilerplate. We no longer need to define `is_valid_target()` and `valid_target_types()` because the superclass gives us those.
   * This all still works with MyPy. We could probably remove even more boilerplate by putting more logic into the superclass, e.g. `create()`, but MyPy wouldn't be able to know what each field is, anymore.

I think #1 is stable. I don't think we'll want to change the name again. (But, we can, need be).

#2 is very fluid. I expect for there to be much churn the next week as we encounter new real-world use cases.

Reply:
Yeah - I think #2 can be completely ditched given a `@dataclass`, the meaning of `Optional` and a `Target` instance. Time will tell.